### PR TITLE
feat: run inserts in batch and in parallel when processing new block

### PR DESCRIPTION
**Pull request by bestmike007**
## Description

This PR improves the new block process improvements by means of running inserts in batch and in parallel.

Previously when the event server process a new block, it roughly process less than 10 transactions per second.
When tested with concurrency=100, it processes more than 30 transactions per second.

## Type of Change

Performance improvements.

## Does this introduce a breaking change?

No.

## Are documentation updates required?

No.

## Testing information

No changes required.

## Checklist
- [x] Code is commented where needed
- [x] Unit test coverage for new or modified code paths
- [x] `npm run test` passes
- [ ] Changelog is updated
- [x] Tag 1 of @rafaelcr or @zone117x for review

### CODE REVIEW

**Comment by bestmike007** [comment]
@rafaelcr @zone117x please help me review this.

File: src/datastore/pg-write-store.ts [hunk 1]
```diff
@@ -256,20 +255,30 @@ export class PgWriteStore extends PgStore {
...
-          await this.updatePoxSyntheticEvents(sql, entry.tx, 'pox3_events', entry.pox3Events);
-          await this.updatePoxSyntheticEvents(sql, entry.tx, 'pox4_events', entry.pox4Events);
-          await this.updateStxLockEvents(sql, entry.tx, entry.stxLockEvents);
-          await this.updateFtEvents(sql, entry.tx, entry.ftEvents);
-          await this.updateNftEvents(sql, entry.tx, entry.nftEvents);
-          await this.updateSmartContracts(sql, entry.tx, entry.smartContracts);
-          await this.updateNamespaces(sql, entry.tx, entry.namespaces);
-          await this.updateNames(sql, entry.tx, entry.names);
+        if (batchedTxData.length > 0) {
+          const q = new PQueue({ concurrency: 4 });
```
Comment (rafaelcr) [inline src/datastore/pg-write-store.ts:259]: Can you make `concurrency` configurable via ENV? Or why is it set to 4? Curious if you did any benchmarks or how did you get to that number

Comment (rafaelcr) [reply src/datastore/pg-write-store.ts:259]: Also, can you initialize the `PQueue` outside this block so we can reuse it across blocks? Will save CPU cycles especially during event replays

Comment (bestmike007) [reply src/datastore/pg-write-store.ts]: 4 is better than 1, and a default value not being too radical if someone is running postgres instance on non-SSD disk.

Added the env entry, please help me review the wording.

File: .env [hunk 2]
```diff
@@ -43,6 +43,10 @@ PG_APPLICATION_NAME=stacks-blockchain-api
 # Limit to how many concurrent connections can be created, defaults to 10
 # PG_CONNECTION_POOL_MAX=10
 
+# Insert concurrency when processing new blocks
+# It'd better set it higher, e.g. 8 or 16, if your postgres is running on SSD and multiple CPU cores
+# INSERT_CONCURRENCY=4
```
Comment (rafaelcr) [inline .env:48]: Please rename this to `STACKS_BLOCK_DATA_INSERT_CONCURRENCY`, also add a comment stating that this value should be set in harmony with `PG_CONNECTION_POOL_MAX` so all the concurrent writes have enough available connections to service them.

File: src/datastore/pg-write-store.ts (hunk 1 above)
Comment (bestmike007) [reply src/datastore/pg-write-store.ts]: > initialize the PQueue outside this block

It's not safe, `await q.onIdle()` will wait for all tasks in the queue, which might come from another process.

Besides, the queue initialization overhead is small compares to all the tasks during event replays.

File: .env (hunk 2 above)
Comment (bestmike007) [reply .env:48]: It doesn't relate to connection pool size, the batch inserts happens within a transaction, which uses a single connection. Batching reduces interaction rounds, and sending requests in parallel reduces time to wait for responses of other irrelevant requests.

Comment (bestmike007) [reply .env]: Renamed.

File: src/datastore/pg-write-store.ts [hunk 3]
```diff
@@ -746,14 +759,35 @@ export class PgWriteStore extends PgStore {
...
+        case 'pox3_events':
+          assert('pox3Events' in entry);
+          events = entry.pox3Events;
+        case 'pox4_events':
+          assert('pox4Events' in entry);
+          events = entry.pox4Events;
+      }
+      if (events == null) {
+        throw new Error(`unknown pox table: ${poxTable}`);
+      }
```
Comment (zone117x) [inline src/datastore/pg-write-store.ts:788]: Fix switch-case fall-through and simplify types

```suggestion
  async updatePoxSyntheticEvents(
    sql: PgSqlClient,
    poxTable: PoxSyntheticEventTable,
    entries: DataStoreTxEventData[]
  ) {
    const values: PoxSyntheticEventInsertValues[] = [];
    for (const entry of entries) {
      let events: DbPoxSyntheticEvent[];
      switch (poxTable) {
        case 'pox2_events':
          events = entry.pox2Events;
          break;
        case 'pox3_events':
          events = entry.pox3Events;
          break;
        case 'pox4_events':
          events = entry.pox4Events;
          break;
        default:
          throw new Error(`unknown pox table: ${poxTable}`);
      }
```

Comment (bestmike007) [reply src/datastore/pg-write-store.ts]: The `entries` does not have to be `DataStoreTxEventData`, using typescript type narrowing is better:

image

Fixed switch-case fall-through.

Comment (zone117x) [reply src/datastore/pg-write-store.ts]: The readability of `updatePoxSyntheticEvents`'s function signature is questionable, although it is type safer for callers, so I'm okay with it 👍

File: src/datastore/pg-write-store.ts (hunk 1 above)
Comment (rafaelcr) [reply src/datastore/pg-write-store.ts]: We process blocks in series so they would not overlap across blocks, but I think we can start with this approach as we test this PR in an env.

File: .env (hunk 2 above)
Comment (rafaelcr) [reply .env]: Ah, you're right, I was thinking from the read side of the API. New name is good.

**Review by rafaelcr — APPROVED** [review]
This looks good to me. We'd like to cut a pre-release Docker image with this PR and run a few tests so we can measure how it performs in the kubernetes pods, especially when interacting with postgres over a network connection.

I'll create a `beta` branch and target your PR into it so we can merge there and we can run tests on our end @bestmike007 , sound good? cc @zone117x

**Review by zone117x — APPROVED** [review]
Thanks a lot for this contribution!

**Comment by bestmike007** [comment]
It'd be better if you run with different concurrency values, and then decide the default value.

**Pull request merged**